### ISSUE TITLE
Training of S4M model

### ISSUE THREAD

**Issue by ndzhizhin** [JusperLee/S4M#4]
Hello!

I am currently trying to train S4M model. Can you provide more information about the training process of S4M? Was it trained with the same config and hyperparameters as TDANet? Did you use the same training code?

Also, could you provide any pretrained checkpoint for tests? Or at least some enchanced wav files with trained S4M? I just want to check the potential quality of the trained model

Hope to hear from you soon!

**Comment by ndzhizhin** [comment]
P.S.
I already tried to train S4M model, but I have a pretty poor quality (both enchanced tracks sounds similar between each other and pretty similar to initial mix wav). Perhaps, you have any tips for training process that could help me

**Comment by JusperLee** [comment]
First, you need to confirm which dataset you are using? Also the sampling rate may affect the performance of the model, please check if your data sampling rate and the model are a match.

Also, you can provide your trainingloss and test result curves in order to consider what is causing the poor performance.

Finally, this project is so old that I can no longer find the corresponding checkpoint.

**Comment by ndzhizhin** [comment]
I am using the LSR2 dataset mentioned in the TDANet repository (https://github.com/JusperLee/TDANet). The sampling rate of the data is 16 kHz. I don't think the issue lies with the data, because I tried training TDANet with the same data, trainer and config, and I got much better results. 

My training loss for S4M and TDANet are similar; it uses pairwise_neg_snr (default in the config) for training, and for validation, I use pairwise_neg_sisdr (also the default).

Full config for S4M:
```
audionet:
  audionet_name: TDANet
  audionet_config:
    out_channels: 128
    in_channels: 512
    num_blocks: 16
    upsampling_depth: 4
    enc_kernel_size: 21 # ms
    num_sources: 2

#Basically, I didnt change audionet_name because I initialize my S4M model in train file with default parameters instead of TDANet 

# Loss config
loss:
  train:
    loss_func: PITLossWrapper
    sdr_type: pairwise_neg_snr
    config:
      pit_from: pw_mtx
      threshold_byloss: true
  val:
    loss_func: PITLossWrapper
    sdr_type: pairwise_neg_sisdr
    config:
      pit_from: pw_mtx
      threshold_byloss: false

# Training config
training:
  system: AudioLightningModule
  gpus: [0]
  parallel: ddp
  epochs: 60
  early_stop:
    monitor: val_loss/dataloader_idx_0
    mode: min
    patience: 5
    verbose: true
  SpeedAug: False
  
# Optim config
optimizer:
  optim_name: adamw
  lr: 0.001
  weight_decay: 0

# Sche config
scheduler: 
  sche_name: ReduceLROnPlateau
  sche_config:
    patience: 2
    factor: 0.5

# Data config
datamodule:
  data_name: LRS2DataModule
  data_config:
    train_dir: F:\lrs2\lrs2_rebuild\audio\wav16k\min\tr
    valid_dir: F:\lrs2\lrs2_rebuild\audio\wav16k\min\cv
    test_dir: F:\lrs2\lrs2_rebuild\audio\wav16k\min\tt
    n_src: 2
    sample_rate: 16000
    segment: 2.0
    normalize_audio: false
    batch_size: 4
    num_workers: 1
    pin_memory: true
    persistent_workers: false

exp:
  exp_name: S4M_06_02
```
Here is result curves with WbandLogger:

1) ![Image](https://github.com/user-attachments/assets/291bd0f6-17b4-412a-a33b-3b03e54d10fb)
2) ![Image](https://github.com/user-attachments/assets/9fdbfd90-8b0a-4496-bb60-61d174a6e596)

As you can see, S4M has a trend, and its loss is slowly decreasing. However, TDANet achieves much better results with a smaller number of epochs, and the enhanced WAVs sound significantly better. At that time, the S4M model latest checkpoint has pretty poor quality (both enhanced tracks sound similar to each other and quite similar to the initial mixed WAV).

**Comment by ndzhizhin** [comment]
Also, do you have any new projects on this topic?

Hope to hear from you soon!
Best regards, Nikita

**Comment by JusperLee** [comment]
maybe you need change

    model = S4M(
        out_channels=128,
        in_channels=512,
        num_blocks=16,
        upsampling_depth=4,
        enc_kernel_size=4, # 4ms
        num_sources=2,
        sample_rate=16000, # or 8000
    )

**Comment by Mashiro009** [comment]
Hi @ndzhizhin, the default setting may be:
```
    model = S4M(
        out_channels=128,
        in_channels=512,
        num_blocks=16,
        upsampling_depth=4,
        enc_kernel_size=4, # 4ms or 2, 2ms
        num_sources=2,
        sample_rate=16000, # or 8000
    )
```
Based on the discussion in issue https://github.com/JusperLee/S4M/issues/5.

**Comment by ndzhizhin** [comment]
Hi!
Thanks, it helped a lot!
Can you answer 2 more questions please:
1) Do you have any new projects on this topic?
2) Have you tried to create a causal version of the S4M model that would work in real time? I have already tried, but the training results are quite bad
